### PR TITLE
Adapt reverse to C++20

### PR DESCRIPTION
**Pull request by gonidelis**

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp [hunk 1]
```diff
@@ -186,65 +324,14 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
     template <typename ExPolicy, typename BidirIter, typename FwdIter,
-        HPX_CONCEPT_REQUIRES_(hpx::traits::is_iterator<BidirIter>::value&&
-                hpx::is_execution_policy<ExPolicy>::value&&
-                    hpx::traits::is_iterator<FwdIter>::value)>
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_iterator<BidirIter>::value &&
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_iterator<FwdIter>::value
+        )>
+    // clang-format on
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:334]: This needs to be deprecated.

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    // clang-format off
+    template <typename Iter,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_input_iterator<Iter>::value ||
+            hpx::traits::is_output_iterator<Iter>::value
+        )>
+    // clang-format on
+    constexpr inline Iter next(Iter first)
+    {
+        return first++;
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:28]:
```suggestion
        return ++first;
```

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        return first + dist;
+    }
+
+    // clang-format off
+    template <typename Iter, typename Sent,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_sentinel_for<Sent, Iter>::value &&
+            hpx::traits::is_input_iterator<Iter>::value ||
+            hpx::traits::is_output_iterator<Iter>::value
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:49]: May I suggest to put parenthesis around the `||` part of the expression to make the intend clear?

Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:49]: @hkaiser So do I go with `is_iterator` or do you prefer the parenthesis workaround?

Comment (hkaiser) [reply libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:49]: Either way is fine, I guess.

**Comment by gonidelis** [comment]
This is more or less ready. I will add a couple of missing ranges docs first thing tomorrow morning and one or two sentinel tests for `reverse_copy`. The main takeaway remains the `ranges::next` facility implementation though.

File: libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx_main.hpp>
+#include <hpx/modules/testing.hpp>
+#include <hpx/parallel/util/ranges_facilities.hpp>
+
+#include "libs/parallelism/algorithms/tests/util/iter_sent.hpp"
```
Comment (gonidelis) [inline libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: @hkaiser  Please also check this moved header.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/detail/advance_to_sentinel.hpp [hunk 5]
```diff
@@ -37,7 +37,7 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
     constexpr inline Iter advance_to_sentinel(Iter first, Sent last)
     {
         return advance_to_sentinel(first, last,
-            typename hpx::traits::is_sized_sentinel_for<Iter, Sent>::type{});
+            typename hpx::traits::is_sized_sentinel_for<Sent, Iter>::type{});
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/detail/advance_to_sentinel.hpp:40]: Thanks again for spotting this!

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        )>
+    // clang-format on
+    constexpr inline Iter next(Iter first, Sent bound)
+    {
+        return hpx::parallel::v1::detail::advance_to_sentinel(first, bound);
+    }
+
+    template <typename Iter, typename Sent>
+    constexpr inline Iter next_(Iter first,
+        typename std::iterator_traits<Iter>::difference_type n, Sent bound)
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:53]: This overload is missing arguments, I believe.

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        while (n > 0 || first != bound)
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:55]: What about `n < 0` (for bidirectional iterators?

**Comment by gonidelis** [comment]
@hkaiser I tweaked the base implementations a little more in order to support sentinels. All tests pass. Once CI is ok, this shall be ready to go. I will unify the multiple `iter_sent` headers into one (or two) along with the full `advance` facility implementation in the upcoming PR.

**Comment by gonidelis** [comment]
@hkaiser  @msimberg  review?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp [hunk 8]
```diff
@@ -41,19 +214,24 @@ namespace hpx { namespace parallel { inline namespace v1 {
             {
             }
 
-            template <typename ExPolicy, typename BidirIter>
-            static BidirIter sequential(
-                ExPolicy, BidirIter first, BidirIter last)
+            template <typename ExPolicy, typename BidirIter, typename Sent>
+            static BidirIter sequential(ExPolicy, BidirIter first, Sent last)
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:218]: This can be marked as `constexpr`

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp [hunk 9]
```diff
@@ -258,3 +360,115 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        {
+            static_assert(
+                (hpx::traits::is_bidirectional_iterator<BidirIter>::value),
+                "Requires at least bidirectional iterator.");
+
+            typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
+
+            return parallel::util::detail::algorithm_result<ExPolicy>::get(
+                hpx::parallel::v1::detail::reverse<BidirIter>().call2(
+                    std::forward<ExPolicy>(policy), is_seq(), first, last));
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:409]: I think you can remove the `is_seq` predicate here and simply rely on the the dispatch facility to figure it out:

```suggestion
            return parallel::util::detail::algorithm_result<ExPolicy>::get(
                hpx::parallel::v1::detail::reverse<BidirIter>().call(
                    std::forward<ExPolicy>(policy), first, last));
```

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/reverse.hpp [hunk 10]
```diff
@@ -113,28 +346,298 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        "instead") typename util::detail::algorithm_result<ExPolicy,
+        typename hpx::traits::range_iterator<Rng>::type>::type
+        reverse(ExPolicy&& policy, Rng&& rng)
+    {
+        typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
+
+        return detail::reverse<
+            typename hpx::traits::range_iterator<Rng>::type>()
+            .call(std::forward<ExPolicy>(policy), is_seq(),
+                hpx::util::begin(rng), hpx::util::end(rng));
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/reverse.hpp:403]: This will most likely not compile anymore (after the changes on master). You need to remove the `is_seq` predicate altogether.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp (hunk 9 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:409]: ok I was going to ask you about `call2`. thanks!

File: libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp (hunk 4 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: Should we move that to <hpx/parallel/util/iter_sent_test_facilities.hpp>? @msimberg any suggestions?

Comment (msimberg) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: I'd like to keep it out of the install if possible. @gonidelis is this file common only to the `libs/parallelism/algorithms` tests? If yes, I'd suggest we do a simple `#include "../../util/iter_sent.hpp"`. Alternatively we add `target_include_directories(${test_name} PRIVATE ../../include)` for all the tests in `libs/parallelism/algorithms` and put it under unique directory in `libs/parallelism/tests/include`.

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp (hunk 6 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:53]: @hkaiser  `advance` and `next` consolidation will take palce in the upcomming PR.

File: libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp (hunk 7 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp:55]: I want touch it for now as the PR need to pass. Will take care of it in the upcomming PR.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/reverse.hpp (hunk 10 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/reverse.hpp]: done

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp [hunk 11]
```diff
@@ -134,16 +283,16 @@ namespace hpx { namespace parallel { inline namespace v1 {
         /// \cond NOINTERNAL
 
         // sequential reverse_copy
-        template <typename BidirIt, typename OutIter>
+        template <typename BidirIt, typename Sent, typename OutIter>
         inline util::in_out_result<BidirIt, OutIter> sequential_reverse_copy(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:287]: Not a showstopper - this could be `constexpr` as well.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp [hunk 12]
```diff
@@ -155,24 +304,27 @@ namespace hpx { namespace parallel { inline namespace v1 {
             {
             }
 
-            template <typename ExPolicy, typename BidirIter, typename OutIter>
+            template <typename ExPolicy, typename BidirIter, typename Sent,
+                typename OutIter>
             static util::in_out_result<BidirIter, OutIter> sequential(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/reverse.hpp:309]: Same here: could be marked as `constexpr`.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp [hunk 13]
```diff
@@ -464,7 +463,7 @@ namespace hpx {
         subrange_t<typename hpx::traits::range_iterator<Rng>::type>>::type
     remove_if(ExPolicy&& policy, Rng&& rng, Pred&& pred, Proj&& proj = Proj());
 
-}    // namespace hpx
+}}    // namespace hpx::ranges
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp:466]: Could we change the use of `tag_invoke` below to `tag_invoke_fallback`, please? This could however be done in a separate PR (please turn this comment into a separate ticket if needed).

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove_copy.hpp [hunk 14]
```diff
@@ -568,7 +568,7 @@ namespace hpx {
     ranges::remove_copy_if(ExPolicy&& policy, Rng&& rng, O dest, Pred&& pred,
         Proj&& proj = Proj());
 
-}    // namespace hpx
+}}    // namespace hpx::ranges
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove_copy.hpp:571]: Same here, please convert this to `tag_invoke_fallback` as well.

File: libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp (hunk 4 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: @gonidelis this is the last thing that should be fixed before this PR can be merged. Could you have a look, please?

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp (hunk 13 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove.hpp:466]: @hkaiser I am confused. Which part are you referring to exactly? I do not use `tag_invoke` in this PR at all.

File: libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp (hunk 4 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: @msimberg @hkaiser There are only two tests outside of algorithms that uses this and they are the `libs/core/iterator_support/tests/unit/is_sentinel_for.cpp`/`is_sized_sentinel.hpp` ones. So do you suggest I go with the cmake option?

Comment (msimberg) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: In that case please move the file to the `iterator_support` tests directory (really an appropriate `include/...` subdirectory), and use `target_include_directories` on the tests that need the directory.

Comment (gonidelis) [reply libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:11]: @msimberg over

**Comment by hkaiser** [comment]
@gonidelis what's left to get this in?

File: libs/core/iterator_support/tests/CMakeLists.txt [hunk 15]
```diff
@@ -12,6 +12,9 @@ if(NOT HPX_WITH_TESTS AND HPX_TOP_LEVEL)
 endif()
 
 if(HPX_WITH_TESTS)
+  add_library(iter_sent_tests INTERFACE)
```
Comment (msimberg) [inline libs/core/iterator_support/tests/CMakeLists.txt:15]: Maybe:
```suggestion
  add_library(hpx_iterator_support_test_utilities INTERFACE)
```
?

File: libs/core/iterator_support/tests/unit/is_sentinel_for.cpp [hunk 16]
```diff
@@ -21,15 +23,15 @@ void is_sentinel_for()
     HPX_TEST_MSG(
         (hpx::traits::is_sentinel_for<std::int64_t, std::int64_t>::value ==
             false),
-        "Integer - integer pair is ncompatible pair");
+        "Integer - integer pair is incompatible pair");
```
Comment (msimberg) [inline libs/core/iterator_support/tests/unit/is_sentinel_for.cpp:26]: Not really part of this PR, but isn't
```
HPX_TEST_MSG(
        !hpx::traits::is_sentinel_for<std::int64_t, std::int64_t>::value,
        "Integer - integer pair is incompatible pair");
```
more idiomatic?

File: libs/core/iterator_support/tests/unit/is_sized_sentinel_for.cpp [hunk 17]
```diff
@@ -4,9 +4,11 @@
 //  Distributed under the Boost Software License, Version 1.0. (See accompanying
 //  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
+#include <hpx/hpx_main.hpp>
```
Comment (msimberg) [inline libs/core/iterator_support/tests/unit/is_sized_sentinel_for.cpp:7]: Why?

File: libs/core/iterator_support/tests/unit/is_sentinel_for.cpp [hunk 18]
```diff
@@ -4,9 +4,11 @@
 //  Distributed under the Boost Software License, Version 1.0. (See accompanying
 //  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
+#include <hpx/hpx_main.hpp>
```
Comment (msimberg) [inline libs/core/iterator_support/tests/unit/is_sentinel_for.cpp:7]: Why does this include hpx_main.hpp?

File: libs/full/include/include/hpx/algorithm.hpp [hunk 19]
```diff
@@ -20,8 +20,6 @@ namespace hpx {
     using hpx::parallel::minmax_element;
     using hpx::parallel::partition;
     using hpx::parallel::partition_copy;
-    using hpx::parallel::reverse;
-    using hpx::parallel::reverse_copy;
     using hpx::parallel::rotate;
```
Comment (msimberg) [inline libs/full/include/include/hpx/algorithm.hpp:23]: Don't forget to update the docs as well.

File: libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp [hunk 20]
```diff
@@ -0,0 +1,40 @@
+//  Copyright (c) 2021 Giannis Gonidelis
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx_main.hpp>
```
Comment (msimberg) [inline libs/parallelism/algorithms/tests/regressions/ranges_facilities.cpp:7]: Is this necessary?

File: libs/parallelism/algorithms/tests/unit/container_algorithms/distance.cpp [hunk 21]
```diff
@@ -4,13 +4,14 @@
 //  Distributed under the Boost Software License, Version 1.0. (See accompanying
 //  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
+#include <hpx/hpx_main.hpp>
```
Comment (msimberg) [inline libs/parallelism/algorithms/tests/unit/container_algorithms/distance.cpp:7]: Is this necessary?

**Review by msimberg — CHANGES_REQUESTED** [review]

File: libs/parallelism/algorithms/tests/unit/container_algorithms/distance.cpp (hunk 21 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/distance.cpp]: You were right, I fixed it.

File: libs/core/iterator_support/tests/unit/is_sentinel_for.cpp (hunk 16 above)
Comment (gonidelis) [reply libs/core/iterator_support/tests/unit/is_sentinel_for.cpp:23]: Done

File: libs/core/iterator_support/tests/CMakeLists.txt (hunk 15 above)
Comment (gonidelis) [reply libs/core/iterator_support/tests/CMakeLists.txt]: Done

**Review by hkaiser — APPROVED** [review]
This is good to go now, thanks!

**Comment by msimberg** [comment]
Thanks @gonidelis!
